### PR TITLE
Pre-processing dataset energies

### PR DESCRIPTION
**Pull request by wiederm**
## Description

It is common in training neural network potentials to perform pre-processing steps on the energies used for training. 
Two common strategies are (described in more detail here https://doi.org/10.1063/5.0160326):
- offsetting the molecular energies by the atomize energies (self energies of the free atoms in vacuum) to avoid numerical instability with the typical considerable total energies
- scaling the range of energies to a unit interval

Both pre-processing strategies require information passed from the dataset to the model, since the transformations must be reversed to obtain total energies.

This PR adds functionality to the dataset class to offset the total energies using atomic self-energies, calculated or provided for a specific level of theory and dataset, as well as the mean and variance of the total or pre-processed energies.

```python
 from modelforge.dataset.qm9 import QM9Dataset
 from modelforge.dataset.utils import FirstComeFirstServeSplittingStrategy

 data = QM9Dataset(for_unit_testing=True)
 dataset = TorchDataModule(
    data, batch_size=8, split=FirstComeFirstServeSplittingStrategy()
    )

# self energy is calculated and removed in prepare_data if `remove_self_energies` is True
dataset.prepare_data(remove_self_energies=True, normalize=False)
# alternatively, if we know the self energies for a dataset it can be provided by passing a dictionary with the atomic number as key a the self-energies as corresponding values
self_energies : {1 : -10., 6 : -1000}
dataset.prepare_data(remove_self_energies=True, normalize=False, self_energies=self_energies)
 ```

### Self-energies

The self-energies are calculated using a linear regression fit if self-energies aren't provided. 
If total energies are required, the self-energies must be passed to the corresponding postprocessing module (`AddSelfEnergies`) to reverse the transformation.

```python
model = SchNET(
    embedding_module=embedding,
    nr_interaction_blocks=nr_interaction_blocks,
    radial_basis_module=rbf,
    cutoff_module=cutoff,
    nr_filters=nr_filters,
    postprocessing=PostprocessingPipeline(
        [AddSelfEnergies(dataset.dataset_statistics)]
    ),
)
```
 
### Normalization

Energies can be normalized by removing the average and dividing by the standard deviation.
This operation can be requested by calling
```python
# self energy is calculated and removed in prepare_data if `remove_self_energies` is True
dataset.prepare_data(remove_self_energies=True, normalize=True)
```

To obtain the total energy both the substraction of the self-energies and the normalization need to be reverted. This can be done by passing both Postprocessing classes to the `ProstprocessingPipeline`:
```python
model = SchNET(
    embedding_module=embedding,
    nr_interaction_blocks=nr_interaction_blocks,
    radial_basis_module=rbf,
    cutoff_module=cutoff,
    nr_filters=nr_filters,
    postprocessing=PostprocessingPipeline(
        [AddSelfEnergies(dataset.dataset_statistics), UndoNormalization(dataset.dataset_statistics)]
    ),
)
```
All of the Postprocessing classes can also be passed without  `dataset.dataset_statistics`, in that case they will check if the dataset_statistics are already provided by the model (a trained model will provide these), and if not, raise an error.

## Todos
  - [x] self-energies infrastructure
  - [x] postprocessing classes 
  - [x] Update LightningDataset and TorchDataset to remove atomic offset/normalize
 
## Status
- [x] Ready to go

### CODE REVIEW

**Comment by jchodera** [comment]
Could adding the offset to the energy degrade the precision of rh energy and hence the precision of our loss function, where we need to subtract two large energies and are looking for << 1 kcal/mol errors in energy differences between snapshots?

If the energies are always converted to double precision before

**Comment by wiederm** [comment]
I agree with that concern. I am unsure about the best strategies, and I think we will have to empirically evaluate this (even if we use double precision).

My plan is to implement the transformation in this PR so that we can control the transformations independently in the computation graph. That will enable us to test the two scenarios: train on energies with and without offset. 

Another possible scenario will be to test whether we want to normalize offset energies or not, again, something that we will have to determine empirically.

**Comment by wiederm** [comment]
@yuanqing-wang , do you have an opionion on that?

**Comment by jchodera** [comment]
Just as a simple example of the issue: A druglike molecule might have a formation energy of around -1e6 kcal/mol. The smallest representable difference that float32 can represent is around 0.0625 kcal/mol if subtracting numbers of this magnitude. Unless promoted to double, this might also lead to lots of noise in gradients.

Bigger molecules (or molecules with heavier atoms, like halides) will have even more trouble.

**Comment by wiederm** [comment]
@jchodera , I think removing the self energies will improve this situation considerably.

**Comment by wiederm** [comment]
@chrisiacovella , did you already implement the self-energies for QM9 somewhere, or do you have opinions on where these should go?
We can also discuss this offline later.

**Comment by chrisiacovella** [comment]
> @chrisiacovella , did you already implement the self-energies for QM9 somewhere, or do you have opinions on where these should go? We can also discuss this offline later.

In the QM9 Dataset, they provided in an SI table thermochemical references for each of the atoms.  E.g., here is the data for hydrogen: 

>             "H": {
>                 "ZPVE": 0.000000 * unit.hartree,
>                 "U_0K": -0.500273 * unit.hartree,
>                 "U_298.15K": -0.498857 * unit.hartree,
>                 "H_298.15K": -0.497912 * unit.hartree,
>                 "G_298.15K": -0.510927 * unit.hartree,
>                 "Cv": 2.981 * unit.calorie_per_mole / unit.kelvin,

For each molecule, I calculate the reference values (i.e., just summing up all the individual atomic contributions) for each of the states/quantities (so reference energy at 0K, at 298.15K, ethanlpy at 298.15K, etc.) 

Obviously, the most relevant for our needs is the 0K energy.  The datafile includes: 

- `internal_energy_at_0K`
- `reference_energy_at_0K`
- `formation_energy_at_0K`

where the `formation_energy_at_0K = internal_energy_at_0K - reference_energy_at_0K`

In the spice_openff dataset, we also have:

- `dft_total_energy`
- `reference_energy`
- `formation_energy`
where again, the `formation_energy  = dft_total_energy - reference_energy`

The non openff level of theory spice has the formation_energy already in the hdf5 they distributed (but not the reference energy for each molecule); I used the same code that was in the downloader script used to generate the hdf5 file Peter posted on zenodo, so those values should be identical (and could be added in trivially, if we aren't going to use the formation_energy directly from the file).  

ANI1x didn't provide the references that I could find in the publication (they may be in the code base but I haven't located them yet).  The same level of theory is used as qm9, so those would probably be reasonable values to use.

File: modelforge/tests/helper_functions.py [hunk 1]
```diff
@@ -130,7 +123,6 @@ def initialize_dataset(
     data = dataset(for_unit_testing=for_unit_testing)
     data_module = TorchDataModule(data, split_file=split_file)
     data_module.prepare_data()
-    data_module.setup(stage=mode)
     return data_module
```
Comment (chrisiacovella) [inline modelforge/tests/helper_functions.py:128]: I think a renaming attempt went rouge: 

`preparePairlist_for_single_batch`

should be `prepare_pairlist_for_single_batch`.

File: modelforge/tests/test_utils.py [hunk 2]
```diff
@@ -53,8 +53,8 @@ def test_rbf(RBF):
 
     from .helper_functions import preparePairlist_for_single_batch, return_single_batch
```
Comment (chrisiacovella) [inline modelforge/tests/test_utils.py:54]: I think the original function needs to be renamed `prepare_pairlist_for_single_batch` and this import to be updated accordingly.

**Review by chrisiacovella — APPROVED** [review]
I fixed one of the issues that was causing CI to fail (weird naming issue) and added in a test for the Welford estimator.  

Most of the other comments are pretty minor things so should just be a quick update and be fine. 

A few things to discuss though in terms of future stuff:
- If we use the linear regression to calculate the self-energies, we need a way to save these to disk, and also probably need a simple way to set them in the code.  Recalculating them is cheap in the overall scheme, but we might want to have a fixed set of values to use with different "tests" (where tests might change the dataset and thus would change the fitted values).  Might be as simple as having "prepare_data" take a dictionary of self-energies.  
-  For many of the datasets we already have the self-energies and and formation energies. Obviously these can just be read from the datasets when we load the hdf5 file (just changing the dictionary entry we want for energy).  I'm guessing we could probably set these as options within the dataset loader itself (`use_formation_energy`), which would have to toggle some internal variables to know that we should not remove self-energies a second time (and raise some error or just warn that we didn't shift a second time if a user specifies it).    Would we want to implement a post processing step to add energy into the model (in which case, we would have to give a way to provide the energy of the atoms in vacuum...either through a supplemental file or by adding this as a group to the hdf5 file...).

File: modelforge/tests/test_utils.py (hunk 2 above)
Comment (wiederm) [reply modelforge/tests/test_utils.py]: thanks for fixing this!

File: modelforge/dataset/dataset.py [hunk 3]
```diff
@@ -476,67 +508,265 @@ def __init__(
...
+
+        # Define the collate function based on the `collate` parameter
+        collate_fn = collate_conformers if collate else None
+
+        # Initialize variables to hold data for regression
+        batch_size = 64
+        # Determine the size of the counts tensor
+        num_molecules = dataset.n_records
+        # Determine up to which Z we detect elements
+        max_atomic_number = 100
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py:554]: I agree that this might be a source of future complications. One way to solve this is to calculate max Z from the dataset that is loaded.

File: modelforge/dataset/dataset.py [hunk 4]
```diff
@@ -476,67 +508,265 @@ def __init__(
...
+        self.split_idxs = None
+        self.train_dataset = None
+        self.test_dataset = None
+        self.val_dataset = None
+        self.transform = transform
+        self.split = split
+        self.split_file = split_file
+        self.dataset_statistics = {}
+
+    @classmethod
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py:525]: I wanted to use this method independent of the ininitalized class --- but now that I revisit this design decision I believe it is better to move the functionality to a function and call this function with a normal instance method. What do you think?

File: modelforge/potential/postprocessing.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+        # Calculate self energies for each atom and then aggregate them per molecule
+        atom_self_energies = self_energies_tensor[atomic_numbers]
+        molecule_self_energies = torch.zeros_like(energies)
+
+        for i, energy in enumerate(energies):
+            # Find atoms belonging to the current molecule
+            mask = molecule_indices == i
+            # Sum self energies for these atoms
+            molecule_self_energies[i] = atom_self_energies[mask].sum()
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/potential/postprocessing.py:110]: Here, we are computing the total self energies depending on the elemental composition of a given molecule.

There are three possible scenarios:
  (1) in training, we have seens this molecule before and know its total self energy. In that case, we could store it.
  (2) in testing, we have *not* seen this molecule before, and the total self energy is unknow
  (3) a specific inference scenario (namly, MD), the molecular self energy is always the same.

The current workflow is appropriate for (2) and (1). There is no other way to handel (2), but also for (1) it seems inefficient to store and load the molecular self energies.I think for (3) the point you are raising is valid, but, I'd argue that we don't want to add the molecular self energies there, so it should still be fine.

File: modelforge/tests/test_training.py [hunk 6]
```diff
@@ -45,3 +45,55 @@ def test_train_with_lightning(dataset: Type[BaseNNP], model_class: Type[BaseNNP]
     model = model.to(torch.float32)
     # Run training loop and validate
     trainer.fit(model, dataset.train_dataloader(), dataset.val_dataloader())
+
+
+def test_pt_lightning():
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/tests/test_training.py:50]: Done!

**Comment by wiederm** [comment]
To pick up the conversation from here: https://github.com/choderalab/modelforge/pull/63#pullrequestreview-1898911989

Currently, each model contains the used atomic self energies as model instance parameters, so these are saved whenever the model is saved. Also, we are writing a toml file containing them.  

We accept the self energies of the datasets (in fact, these should be preferred) as dict here: https://github.com/choderalab/modelforge/blob/c17db775c648098adfc4f75cf9d4eaec2bbc9cb2/modelforge/dataset/dataset.py#L557
We should provide a method to obtain them for the dataset internally though. I will open an issue about this.

File: modelforge/dataset/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py]: Sorry, let me clarify this comment: I don't mean the torch dataset, I mean the internal representation in the h5 file.

File: modelforge/tests/test_dataset.py [hunk 7]
```diff
@@ -288,11 +304,36 @@ def test_numpy_dataset_assignment(dataset):
...
+        data, batch_size=32, split=FirstComeFirstServeSplittingStrategy()
+    )
+
+    # self energy is calculated and removed in prepare_data if `remove_self_energies` is True
+    dataset.prepare_data(remove_self_energies=True, normalize=True)
+
+    assert dataset.dataset_statistics
+    self_energies = dataset.dataset_statistics["self_energies"]
+    # only 4 elements present in the reduced QM9 dataset
+    assert len(self_energies) == 4
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/tests/test_dataset.py:335]: I am not completely sure why these are off actually, they matched the last time I checked this. I will open an issue about this.

**Comment by wiederm** [comment]
I will spin off some of the comments in new issues so we can merge this PR as soon as possible.
New issues:
https://github.com/choderalab/modelforge/issues/72
https://github.com/choderalab/modelforge/issues/73

**Comment by wiederm** [comment]
And, as @chrisiacovella mentioned [here](https://github.com/choderalab/modelforge/pull/63#pullrequestreview-1898911989), I have added a reference_ase dictionary for the QM9 dataset that we can use internally.

**Comment by wiederm** [comment]
I added a `AtomicSelfEnergies` dataclass, that takes a dictionary of {str, float} atomic self-energies and returns the atomic self energies both by element name and atomic index. I added it to the qm9 dataclass.

File: modelforge/tests/test_dataset.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply modelforge/tests/test_dataset.py:335]: I wouldn't say these are really all that off.  All but hydrogen have less than a 1% difference.  Hydrogen actually has the smallest numerial difference (delta), but because it is nearly 2 orders of magnitude smaller, it just seems more significant.  

```
element |  DFT              | regressed         | delta         | % diff
H       |   -1313.466862    |   -1585.320503    |  271.8536413  |  20.69741151
C       |  -99366.70746     |  -99963.18962     |  596.4821665  |   0.600283719
N       | -143309.938       | -143744.2422      |  434.3042769  |   0.303052449
O       | -197082.0672      | -197504.4753      |  422.4081034  |   0.21433107
F       | -261811.5456      | -262199.961       |  388.415491   |   0.148356899
```

But since the regressed values are all consistently higher than the DFT values, if we use them we are going to end up with positive values for small molecules and much smaller energies for large molecules.  A few quick calculations of the formation energy. 

Methane formation energy
DFT self energy: -1656.8639 kJ/mol
regressed self energy: 27.0328 kJ/mol

Ammonia formation energy:
DFT self energy: -1158 kJ/mol
regressed self energy: 91 kJ/mol

water: 
DFT self energy: -891 kJ/mol
regressed self energy: 74.55 kJ/mol

acetylene: 
DFT: -1612 kJ/mol
regressed: 123 kJ/mol

CCC(=O)C=O
DFT: -4922 KJ/mol
regressed: -60 KJ/mol

The larger question: does this actually matter if we can just add the same constant self energy values back on later?

File: modelforge/potential/postprocessing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply modelforge/potential/postprocessing.py:110]: That explanation seems reasonable. I don't think it will add much overhead to the training to recompute.

File: modelforge/dataset/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply modelforge/dataset/dataset.py]: I guess the complication of having max_atomic_number be a variable is that a user could under-allocate (we would need to do a check as we read in the batches to make sure haven't under allocated).  It is probably fine to just leave it as is, since I doubt we are going to be trending into the synthetic elements anytime soon, for obvious reasons.  And again, since this didn't seem to in anyway impact the speed of the calculation, I think it's fine to just pad out to 100 for safety.

File: modelforge/dataset/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply modelforge/dataset/dataset.py]: I agree it should probably be normal class function.

File: modelforge/dataset/dataset.py [hunk 8]
```diff
@@ -476,67 +508,265 @@ def __init__(
...
+        Dict[int, float]
+            A dictionary mapping atomic numbers to their calculated self energies.
+        """
+        log.info("Computing self energies for elements in the dataset.")
+
+        # Define the collate function based on the `collate` parameter
+        collate_fn = collate_conformers if collate else None
+
+        # Initialize variables to hold data for regression
+        batch_size = 64
```
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply modelforge/dataset/dataset.py]: I'm guessing this wouldn't be set directly by a user, but rather set explicitly when called by prepare data.  My thought for why this should be modifiable is from past experience having to modify codes because, e.g. character buffers were not set sufficiently high enough and there was no way to easily change it. More thinking that buffer sizes might be something that we may want to up as newer hardware comes out (or be determined on the fly based on hardware).   I just, in general dislike hardcoded variables deep in code; it's much easier to have those at the upper levels, even if not directly modifiable (as in, have these be set in the function that calls it).

File: modelforge/dataset/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py]: I changed this and these are now functions.

File: modelforge/dataset/dataset.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py]: I understand. I will expose this in the upper levels and make it user modifiable.

File: modelforge/tests/test_dataset.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/tests/test_dataset.py]: I believe this doesn't matter!
